bundles/org.openhab.binding.melcloud/src/main/java/org/openhab/binding/melcloud/internal/api/MelCloudConnection.java

```diff
@@ -0,0 +1,215 @@
+                        if (floor.getDevices() != null) {
+                            devices.addAll(floor.getDevices());
+                        }
+                        floor.getAreas().forEach(area -> {
+                            if (area.getDevices() != null) {
+                                devices.addAll(area.getDevices());
+                            }
+                        });
+                    });
+                });
+                logger.debug("Found {} devices", devices.size());
+
+                return devices;
+            } catch (IOException | JsonSyntaxException e) {
+                setConnected(false);
+                throw new MelCloudCommException("Error occured during device list poll", e);
+            }
+        }
+        throw new MelCloudCommException("Not connected to MELCloud");
+    }
+
+    public DeviceStatus fetchDeviceStatus(int deviceId, int buildingId) throws MelCloudCommException {
+        if (isConnected()) {
+            String url = DEVICE_URL + String.format("/Get?id=%d&buildingID=%d", deviceId, buildingId);
+            try {
+                String response = HttpUtil.executeUrl("GET", url, getHeaderProperties(), null, null,
+                        TIMEOUT_MILLISECONDS);
+                logger.debug("Device status response: {}", response);
+                DeviceStatus deviceStatus = gson.fromJson(response, DeviceStatus.class);
+                return deviceStatus;
+            } catch (IOException | JsonSyntaxException e) {
+                setConnected(false);
+                throw new MelCloudCommException("Error occured during device status fetch", e);
+            }
+        }
+        throw new MelCloudCommException("Not connected to MELCloud");
+    }
+
+    public DeviceStatus sendDeviceStatus(DeviceStatus deviceStatus) throws MelCloudCommException {
+        if (isConnected()) {
+            String content = gson.toJson(deviceStatus, DeviceStatus.class);
+            logger.debug("Sending device status: {}", content);
+            InputStream data = new ByteArrayInputStream(content.getBytes(StandardCharsets.UTF_8));
+            try {
+                String response = HttpUtil.executeUrl("POST", DEVICE_URL + "/SetAta", getHeaderProperties(), data,
+                        "application/json", TIMEOUT_MILLISECONDS);
+                logger.debug("Device status sending response: {}", response);
+                return gson.fromJson(response, DeviceStatus.class);
+            } catch (IOException | JsonSyntaxException e) {
+                setConnected(false);
+                throw new MelCloudCommException("Error occured during device command sending", e);
+            }
+        }
+        throw new MelCloudCommException("Not connected to MELCloud");
+    }
+
+    public HeatpumpDeviceStatus fetchHeatpumpDeviceStatus(int deviceId, int buildingId) throws MelCloudCommException {
+        if (isConnected()) {
```

Comment:
You could (for a next time) get rid of this repeated pattern, by calling a method like:

```java
private void assertConnected() {
    if (!isConnected) {
        throw new MelCloudCommException("Not connected to MELCloud");
     }
} 
```

The benefit is that you don't end up with nesting all your method bodies. You could also just start with the fragment above and put the body outside of the brackets but then you still have the repetition.

Although personally I'm in doubt whether throwing an exception where there is no connection is maybe too much.